mpmath/libmp/libmpf.py

```diff
@@ -1323,41 +1375,63 @@ def from_str(x, prec, rnd=round_fast):
     if x in special_str:
         return special_str[x]
 
+    if base is None:
+        if x.startswith('0b') or x.startswith('-0b'):
+            base = 2
+        elif x.startswith('0x') or x.startswith('-0x'):
+            base = 16
+        elif x.startswith('0o') or x.startswith('-0o'):
+            base = 8
+        else:
+            base = 10
+
     if '/' in x:
         p, q = x.split('/')
         p, q = p.rstrip('l'), q.rstrip('l')
-        return from_rational(int(p), int(q), prec, rnd)
+        return from_rational(int(p, base), int(q, base), prec, rnd)
 
-    man, exp = str_to_man_exp(x, base=10)
+    man, exp = str_to_man_exp(x, base)
 
-    # XXX: appropriate cutoffs & track direction
-    # note no factors of 5
-    if abs(exp) > 400:
-        s = from_int(man, prec+10)
-        s = mpf_mul(s, mpf_pow_int(ften, exp, prec+10), prec, rnd)
-    else:
-        if exp >= 0:
-            s = from_int(man * 10**exp, prec, rnd)
+    if base == 10:
+        # XXX: appropriate cutoffs & track direction
+        # note no factors of 5
+        if abs(exp) > 400:
+            s = from_int(man, prec+10)
+            s = mpf_mul(s, mpf_pow_int(ften, exp, prec+10), prec, rnd)
         else:
-            s = from_rational(man, 10**-exp, prec, rnd)
+            if exp >= 0:
+                s = from_int(man * 10**exp, prec, rnd)
+            else:
+                s = from_rational(man, 10**-exp, prec, rnd)
+    elif base in base_exp:
+        s = from_man_exp(man, exp * base_exp[base], prec, rnd)
+    else:
+        raise NotImplementedError
     return s
 
-# Binary string conversion. These are currently mainly used for debugging
-# and could use some improvement in the future
+def from_hex(x, prec=None, rnd=round_fast):
+    """Create a raw mpf from a hexadecimal literal, rounding in the
+    specified direction if *prec* is not *None*. The literal syntax is
+    the same as for Python floats.
+    """
+    man, exp = str_to_man_exp(x, base=16)
+    return from_man_exp(man, exp<<2, prec, rnd)
 
-def from_bstr(x):
-    man, exp = str_to_man_exp(x, base=2)
-    man = MPZ(man)
-    sign = 0
-    if man < 0:
-        man = -man
-        sign = 1
-    bc = bitcount(man)
-    return normalize(sign, man, exp, bc, bc, round_floor)
+def from_oct(x, prec=None, rnd=round_fast):
+    """Create a raw mpf from an octal literal, rounding in the
+    specified direction if *prec* is not *None*. The literal syntax is
+    the same as for Python floats.
+    """
+    man, exp = str_to_man_exp(x, base=8)
+    return from_man_exp(man, (exp<<1)+exp, prec, rnd)
 
-def to_bstr(x):
-    sign, man, exp, bc = x
-    return ['','-'][sign] + numeral(man, size=bitcount(man), base=2) + ("e%i" % exp)
+def from_bin(x, prec=None, rnd=round_fast):
```

Comment:
This also breaks gmpy builds.
```suggestion
def from_bin(x, prec=0, rnd=round_fast):
```